feat: reconnect bookingreference for new credential

vijayraghav-io opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #16785 
- Fixes CAL-4398 (Linear issue number - should be visible at the bottom of the GitHub issue description)

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->
**Before Changes:** 
https://www.loom.com/share/a3ec8e7b7181419d91b12abfa9729a99?sid=a8440350-c26e-4778-abd9-73ce539338f6

**After Changes**
https://www.loom.com/share/a36bf823820947d2b0d2225853363885?sid=866df2d7-dafb-4a63-8e72-cd3f5ed8f293

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] - N/A - I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

vijayraghav-io committed:
feat: reconnect bookingreference for new credential

vijayraghav-io committed:
nit: correct error message

keithwillcode commented:
@vijayraghav-io Really appreciate your work on this PR. Helps us move forward in using repositories more.

A suggestion for the future. Avoid big refactors with feature changes.
- If the refactor causes an issue for some reason and we need to revert/rollback, the new feature is also lost in the process. Ideally, the feature is added in 1 PR and the refactor is done in another PR to avoid this situation.
- It makes reviewing the PR harder because the context is wider.

keithwillcode reviewed (changes requested):
There don't seem to be new tests added to verify this functionality. Please add 🙏

vijayraghav-io commented:
> @vijayraghav-io Really appreciate your work on this PR. Helps us move forward in using repositories more.
> 
> A suggestion for the future: avoid big refactors with feature changes.
> - If the refactor causes an issue for some reason and we need to revert/rollback, the new feature is also lost in the process. Ideally, the feature is added in 1 PR and the refactor is done in another PR to avoid this situation.
> - It makes reviewing the PR harder because the context is wider.

Thankyou! :pray @keithwillcode 

Sure, will take care of having refactor in a separate PR. I can split this right away, will update with new PR for refactor.

Also will add tests for the fix.

vijayraghav-io commented:
sorry for the delay, in middle of other PR, will update this PR by this week.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
refactored to have only issue fix

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
nit

vijayraghav-io commented:
creation of tests in progress....

vijayraghav-io commented on packages/lib/server/repository/bookingReference.ts:
```diff
@@ -49,4 +57,98 @@ export class BookingReferenceRepository {
+      }
+
+      //Detect bookingReferences to connect with new Credential.
+      const bookingReferences = await prisma.bookingReference.findMany({
+        where: {
+          type: newCredential.type,
+          booking: {
+            ...(newCredential.userId
+              ? { userId: newCredential.userId }
+              : { userId: { in: teamMembersUserIds } }),
+          },
+          credentialId: null,
+          ...(!newCredential.selectedCalendars || newCredential.selectedCalendars.length === 0
+            ? { externalCalendarId: null } // for non-calendar apps
+            : {
+                externalCalendarId: {
+                  in: newCredential.selectedCalendars
+                    .filter((selectedCalendar) => !!selectedCalendar.externalId)
+                    .map((selectedCalendar) => selectedCalendar.externalId as string),
+                },
+              }),
+        },
+      });
```
raw SQL Query generated -
```
SELECT
    "BookingReference"."id",
    "BookingReference"."type"
FROM
    "public"."BookingReference"
LEFT JOIN
    "public"."Booking" AS "j1" ON "j1"."id" = "BookingReference"."bookingId"
WHERE
    "BookingReference"."type" = $1
    AND "j1"."userId" = $2
    AND "j1"."id" IS NOT NULL
    AND "BookingReference"."credentialId" IS NULL
    AND "BookingReference"."externalCalendarId" IN ($3)
    AND "BookingReference"."deleted" IS NULL
OFFSET $4;
```

vijayraghav-io committed:
nit: select fields in query

vijayraghav-io commented on packages/lib/server/repository/bookingReference.ts:
```diff
@@ -49,4 +57,106 @@ export class BookingReferenceRepository {
+      }
+
+      //Detect bookingReferences to connect with new Credential.
+      const bookingReferences = await prisma.bookingReference.findMany({
+        where: {
+          type: newCredential.type,
+          booking: {
+            ...(newCredential.userId
+              ? { userId: newCredential.userId }
+              : { userId: { in: teamMembersUserIds } }),
+          },
+          credentialId: null,
+          ...(!newCredential.selectedCalendars || newCredential.selectedCalendars.length === 0
+            ? { externalCalendarId: null } // for non-calendar apps
+            : {
+                externalCalendarId: {
+                  in: newCredential.selectedCalendars
+                    .filter((selectedCalendar) => !!selectedCalendar.externalId)
+                    .map((selectedCalendar) => selectedCalendar.externalId as string),
+                },
+              }),
+        },
+        select: {
+          id: true,
+          bookingId: true,
+        },
+      });
```
raw SQL Query generated -
```
SELECT
    "BookingReference"."id",
    "BookingReference"."type"
FROM
    "public"."BookingReference"
LEFT JOIN
    "public"."Booking" AS "j1" ON "j1"."id" = "BookingReference"."bookingId"
WHERE
    "BookingReference"."type" = $1
    AND "j1"."userId" = $2
    AND "j1"."id" IS NOT NULL
    AND "BookingReference"."credentialId" IS NULL
    AND "BookingReference"."externalCalendarId" IN ($3)
    AND "BookingReference"."deleted" IS NULL
OFFSET $4;
```
IMO: results in single left-join, not much complex.

vijayraghav-io committed:
added test for bookingRef reconnect

vijayraghav-io committed:
added test for google_video

vijayraghav-io committed:
nit

vijayraghav-io commented:
Hi @keithwillcode,
Sorry for the delay!
Updated this PR to have issue fix only, without using Repository approach.

Updates in comparison to previous approach :
1. Reverted the use of `CredentialRepository` wherever the Credential is being created with `prisma.credential.create()` as it was before.
2. Calling the `reconnectWithNewCredential()` function after the `credential` is created, wherever it is being created.

IMO the best place to have this function - `reconnectWithNewCredential()` is inside BookingReferenceRepository.

To do:
Will create a new PR that will -
1. Replace all `prisma.credential.create()` with `CredentialRepository.Create()`. 
2. `BookingReferenceRepository.reconnectWithNewCredential()` will be called from within the `CredentialRepository.Create()`. 
This also ensures that in future if, we have a new req to write code to create new `credential` , dev or reviewer need not remember to call `BookingReferenceRepository.reconnectWithNewCredential()` after credential creation. But just ensure to use the CredentialRepository.Create(), instead of directly using prisma. 
3. Calling the `reconnectWithNewCredential()` from outside the Repository, introduced from this PR will have to be removed, after this current PR is merged.

vijayraghav-io commented:
> @vijayraghav-io could pls fix the conflicts🙏

🙏  sure will do

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
updates after merge

vijayraghav-io commented:
@anikdhabal , resolved merge conflicts.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io commented:
Gentle reminder!

keithwillcode commented:
@emrysal Can you please review this? I know you had fixed a lot of the credentials issues with booking references in the db.

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

Devanshusharma2005 reviewed (changes requested):
hey @vijayraghav-io Thanks for the PR! Just noticed there aren’t any new tests to check the "reconnectWithNewCredential" functionality could you add some automated tests to make sure orphaned BookingReferences get reconnected properly for both user and team cases? Also, there are some merge conflicts right now, so please rebase your branch onto main, fix those up, and push the changes. Once you’ve added the tests and sorted the conflicts, just ping the reviewers so we can take another look. Thanks!

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
removing external use of BookingReferenceRepository

vijayraghav-io committed:
removed external usage of BookingReferenceRepository.reconnectWithNewCredential

vijayraghav-io committed:
nit

vijayraghav-io committed:
call 'BookingReferenceRepository.reconnectWithNewCredential' from CredentialRepository.create

vijayraghav-io committed:
call 'BookingReferenceRepository.reconnectWithNewCredential' inside 'SelectedCalendarRepository.create'.

vijayraghav-io committed:
replacing prisma.credential.create with CredentialRepository.create

vijayraghav-io committed:
nit, define BaseCredentialCreateInput

vijayraghav-io committed:
replaced all prisma.credential.create with CredentialRepository.create

vijayraghav-io committed:
reconnect bookingReferences when on selectedCalendar upsert

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
Merge branch 'main' into feat_updateBookingRefs_credCreate

vijayraghav-io commented:
All tests are passing!

anikdhabal commented:
@vijayraghav-io can you pls fix the conflicts?

vijayraghav-io committed:
Merge branch 'main' of https://github.com/calcom/cal.com into feat_updateBookingRefs_credCreate

vijayraghav-io committed:
updates/review comments

vijayraghav-io committed:
update type definition,..

vijayraghav-io committed:
fix for type check

vijayraghav-io committed:
fix for type check

vijayraghav-io commented on packages/lib/server/repository/credential.ts:
```diff
@@ -6,17 +6,24 @@ import { safeCredentialSelect } from "@calcom/prisma/selects/credential";
 
-type CredentialCreateInput = {
+type BaseCredentialCreateInput = {
   type: string;
   key: any;
-  userId: number;
   appId: string;
   delegationCredentialId?: string | null;
+  subscriptionId?: string | null;
+  paymentStatus?: string | null;
+  billingCycleStart?: number | null;
 };
```
This is an existing issue that key is defined as any,
updating this requires lot of code changes which is not related to current PR scope.
